design/ceph/object/swift-and-keystone-integration.md

```diff
@@ -0,0 +1,321 @@
+  letter followed by an underscore. E.g. `urlPrefix` maps to
+  `rgw_swift_url_prefix`. They are optional. If not given, the defaults
+  of the corresponding RGW option apply.
+
+The access to the Swift API is granted by creating a subuser of an RGW
+user. While commonly the access is granted via projects
+mapped from Keystone, explicit creation of subusers is supported by
+extending the `cephobjectstoreuser` resource with a new optional section
+`spec.subUsers`:
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephObjectStoreUser
+metadata:
+  name: my-user
+  namespace: rook-ceph
+spec:
+  store: my-store
+  displayName: my-display-name
+  quotas:
+    maxBuckets: 100
+    maxSize: 10G
+    maxObjects: 10000
+  capabilities:
+    user: "*"
+    bucket: "*"
+  subUsers:
+  - name: swift  [1]
+    access: full [2]
+```
+
+Annotations:
+* `[1]` This is the name of the subuser without the `username:` prefix
+  (see below for more explanation). The `name` must be unique within
+  the `subUsers` list.
+
+* `[2]` The possible values are: `read`, `write`, `readwrite`,
+  `full`. These values take their meanings from the possible values of
+  the `--access-level` option of `radosgw-admin subuser create`, as
+  documented in the [radosgw admin guide](
+  https://docs.ceph.com/en/octopus/radosgw/admin/#create-a-subuser).
+
+* `name` and `access` are required for each item in `subUsers`.
+
+When changing the subuser-configuration in the CR, this is reflected
+on the RGW side:
+* Subsers are deleted and created to match the list of subusers in the
+  resource.
+* If the access level for an existing user is changed no new
+  credentials are created, but the existing credentials are kept.
+* If a subuser is deleted the corresponding credential secret is
+  deleted as well.
+* Changing only the order of the subuser list does not trigger a
+  reconcile.
+
+The subusers are not mapped to a separate CR for the
+following reasons:
+
+* The full subuser names are prefixed with the username like
+  `my-user:subusername`, so being unique within the CR guarantees
+  global uniqueness.
+
+  Unlike `radosgw-admin` the subuser name in the CRD must be provided
+  *without the prefix* (radosgw-admin allows both, to omit or include
+  the prefix).
+
+* The subuser abstraction is very simple – only a name and an access
+  level can be configured, so a separate resource would not be
+  appropriate complexity wise.
+
+Like for the S3 access keys for the users, the swift keys created for the
+sub-users will be automatically injected into Secret objects. The
+credentials for the subusers are mapped to separate secrets, in the
+case of the example the following secret will be created:
+```yaml
+apiVersion:
+kind: Secret
+metadata:
+  name: rook-ceph-object-subuser-my-store-my-user-swift [1]
+  namespace: rook-ceph
+data:
+  SWIFT_USER: my-user:swift                             [2]
+  SWIFT_SECRET_KEY: $KEY                                [3]
```

Comment:
@sebastianriese include the endpoint info as well. The secret for the `my-user` need not to share with the applications using subuser `swift` right?